Add polygons obstacles (#343)
lambdaclass/mirra_backend: apps/arena/docs/collision-detection.md

# Collision detection

> ## Jeffrey thompson implementation

Reviewer: ```suggestion
## Jeffrey Thompson's implementation
```

It is necessary to clarify that there are other types of collisions that we do not need to implement at the moment, for example Polygon/Polygon.

> ## SAT (Simple Axis Theorem) implamentation

Reviewer: ```suggestion
## SAT (Simple Axis Theorem) implementation
```

A collision detection library implementation based on the [SAT theorem](https://dyn4j.org/2010/01/sat/).

> We'll keep working with entities with the same shapes as the Jeffrey thompson

Reviewer: ```suggestion
We'll keep working with entities with the same shapes as the ones used by Jeffrey Thompson.
```

> To detect a collision between two entities we'll iterate over an axis for each pair of vertices from both entities that collided,

Reviewer: ```suggestion
To detect a collision between two entities, we'll iterate over an axis for each pair of vertices from both entities that collided.
```

> for circle shaper we'll check the axis that's formed between the circle center and the closes polygon vertex, and then cast the

Reviewer: ```suggestion
For circular shapes, we'll check the axis that's formed between the circle center and the closest polygon vertex, and then cast the
```

> maximun and minimun vertex to the normal of each pair of vertices. As soon as we find an axis where the cast of the shapes are not overlapping we can safely say that the entities are not colliding

Reviewer: ```suggestion
maximum and minimum vertex to the normal of each pair of vertices. As soon as we find an axis where the cast of the shapes are not overlapping, we can safely say that the entities are not colliding.
```

> #### Tiangle example for axis where the cast in the axis are overlapping

Reviewer: ```suggestion
#### Triangle example for axis where the cast in the axis are overlapping
```

![Polygons axis overlapping](./images/sat-overlapping.jpg "Polygons axis overlapping")

> #### Tiangle example for axis where the cast in the axis are not overlapping

Reviewer: ```suggestion
#### Triangle example for axis where the cast in the axis are not overlapping
```

![Polygons axis not overlapping](./images/sat-no-overlapping.jpg "Polygons axis not overlapping")

> The collisions we curretly support are:

Reviewer: ```suggestion
The collisions we currently support are:
```

1. Circle with polygon:
![Circle/Polygon not colliding](./images/circle-polygon-not-colliding.jpg "Circle/Polygon not colliding")

![Circle/Polygon colliding](./images/circle-polygon-colliding.jpg "Circle/Polygon colliding")

## Collision resolution

> The Jeffrey thompson implementation for collision detection is ideal since it's the cheapest way to 

Reviewer: ```suggestion
The Jeffrey Thompson's implementation for collision detection is ideal since it's the cheapest way to 
```

> detect that two entities collided, but it's incomplete we can only detect that a collision occurred 

Reviewer: ```suggestion
detect that two entities collided, but it's incomplete since we can only detect that a collision occurred 
```

> nothing else, so we decided to take the following approach:

Reviewer: ```suggestion
and nothing else, so we decided to take the following approach:
```

> 1. We use the Jeffrey thompson implementation to detect that a collision actually happened

Reviewer: ```suggestion
1. We use the Jeffrey Thompson's implementation to detect that a collision actually happened.
```

2. Using SAT we can get the normal that will resolve that collision and the amount of movement we should do 

> so the entities are no longer colliding, so when a collision occur we will use sat to push the entities in the direction that, with the minimum movement, will resolve that collision

Reviewer: ```suggestion
so the entities are no longer colliding, so when a collision occurs we will use SAT to push the entities in the direction that, with the minimum movement, will resolve that collision.
```
